File: nodejs/claude/sample-agent/src/token-cache.ts
// ------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// ------------------------------------------------------------------------------

export function createAgenticTokenCacheKey(agentId: string, tenantId?: string): string {
  return tenantId ? `agentic-token-${agentId}-${tenantId}` : `agentic-token-${agentId}`;
}

// A simple example of custom token resolver which will be called by observability SDK when needing tokens for exporting telemetry
export const tokenResolver = (agentId: string, tenantId: string): string | null => {
  try {
    // Use cached agentic token from agent authentication
    const cacheKey = createAgenticTokenCacheKey(agentId, tenantId);
    const cachedToken = tokenCache.get(cacheKey);

    if (cachedToken) {
      return cachedToken;
    } else {
      return null;
    }
  } catch (error) {
    console.error(`❌ Error resolving token for agent ${agentId}, tenant ${tenantId}:`, error);
    return null;
  }
};

/**
 * Simple custom in-memory token cache
 * In production, use a more robust caching solution like Redis
 */
class TokenCache {
  private cache = new Map<string, string>();

  set(key: string, token: string): void {
    this.cache.set(key, token);
    console.log(`🔐 Token cached for key: ${key}`);
  }

  get(key: string): string | null {
    const entry = this.cache.get(key);
    if (!entry) {
      console.log(`🔍 Token cache miss for key: ${key}`);
      return null;
    }
    return entry;
  }

  has(key: string): boolean {
    const entry = this.cache.get(key);
    return !!entry;
  }
}

// Create a singleton instance for the application
const tokenCache = new TokenCache();

export default tokenCache;

File: python/claude/sample-agent/token_cache.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""
Token Cache
Caches agentic tokens for observability export.
"""

import logging

logger = logging.getLogger(__name__)

# In-memory cache for agentic tokens
# Key format: "tenant_id:agent_id"
_token_cache: dict[str, str] = {}


def cache_agentic_token(tenant_id: str, agent_id: str, token: str) -> None:
    """
    Cache an agentic token for later use by observability exporter.

    Args:
        tenant_id: Tenant identifier
        agent_id: Agent identifier
        token: Agentic authentication token
    """
    cache_key = f"{tenant_id}:{agent_id}"
    _token_cache[cache_key] = token
    logger.debug(f"Cached agentic token for {cache_key}")


def get_cached_agentic_token(tenant_id: str, agent_id: str) -> str | None:
    """
    Retrieve a cached agentic token.

    Args:
        tenant_id: Tenant identifier
        agent_id: Agent identifier

    Returns:
        Cached token if found, None otherwise
    """
    cache_key = f"{tenant_id}:{agent_id}"
    token = _token_cache.get(cache_key)

    if token:
        logger.debug(f"Retrieved cached token for {cache_key}")
    else:
        logger.debug(f"No cached token found for {cache_key}")

    return token


def clear_token_cache() -> None:
    """Clear all cached tokens."""
    _token_cache.clear()
    logger.debug("Token cache cleared")
